usecase/src/main/java/org/mobilitydata/gtfsvalidator/usecase/ValidateStopTimeDepartureTimeAfterArrivalTime.java

```diff
@@ -0,0 +1,86 @@
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.mobilitydata.gtfsvalidator.usecase;
+
+import org.apache.logging.log4j.Logger;
+import org.mobilitydata.gtfsvalidator.domain.entity.StopTimeArrivalTimeAfterDepartureTimeNotice;
+import org.mobilitydata.gtfsvalidator.domain.entity.gtfs.stoptimes.StopTime;
+import org.mobilitydata.gtfsvalidator.usecase.port.GtfsDataRepository;
+import org.mobilitydata.gtfsvalidator.usecase.port.ValidationResultRepository;
+import org.mobilitydata.gtfsvalidator.usecase.utils.TimeUtils;
+
+import java.util.Map;
+
+/**
+ * Use case to validate that `departure_time` does not precede the `arrival_time` date if both are fields are
+ * provided. This use case is triggered after completing the {@code GtfsDataRepository} provided in the constructor with
+ * {@code StopTime} entities.
+ */
+public class ValidateStopTimeDepartureTimeAfterArrivalTime {
+    private final GtfsDataRepository dataRepo;
+    private final ValidationResultRepository resultRepo;
+    private final TimeUtils timeUtils;
+    private final Logger logger;
+
+    /**
+     * @param dataRepo   a repository storing the data of a GTFS dataset
+     * @param resultRepo a repository storing information about the validation process
+     * @param logger     a logger used to log information about the validation process
+     */
+    public ValidateStopTimeDepartureTimeAfterArrivalTime(final GtfsDataRepository dataRepo,
+                                                         final ValidationResultRepository resultRepo,
+                                                         final TimeUtils timeUtils,
+                                                         final Logger logger) {
+        this.dataRepo = dataRepo;
+        this.resultRepo = resultRepo;
+        this.timeUtils = timeUtils;
+        this.logger = logger;
+    }
+
+    /**
+     * Use case execution method: checks if for each {@code StopTime} contained in {@link GtfsDataRepository} that
+     * `departure_time` date does not precede the `arrival_time`. If this requirement is not met, a
+     * {@code StopTimeArrivalTimeAfterDepartureTimeNotice} is added to the {@code ValidationResultRepo} provided in the
+     * constructor.
+     * This verification is executed if {@link StopTime} has non-null value for both fields `departure_time` and
+     * `arrival_time`.
+     */
+    public void execute() {
+        logger.info("Validating rule 'E043 - `departure_time` and `arrival_time` out of order");
+        dataRepo.getStopTimeAll().forEach((tripId, innerMap) -> {
+            final Map<Integer, StopTime> tripStopTimes = innerMap;
+            tripStopTimes.forEach((stopSequence, stopTime) -> {
```

Comment:
I'd suggest looking at implementing all the "time comparison" rules within stop_times.txt in the same use case. Looping through all stop_times.txt records for all trips multiple times will definitely have a negative performance impact, and some of the rules are actually part of the same IF/THEN structure.

Reply:
I think we can go ahead like this and will be able to refactor quite easily when we have a better picture of all the use cases, given the limited resources we have at the moment.